// karabiner.ts/dist/index.js
let G = [
  "left_control",
  "left_shift",
  "left_option",
  "left_command",
  "right_control",
  "right_shift",
  "right_option",
  "right_command",
  "fn"
], ve = [...G, "caps_lock"], Me = [
  "return_or_enter",
  "escape",
  "delete_or_backspace",
  "delete_forward",
  "tab",
  "spacebar",
  "hyphen",
  "equal_sign",
  "open_bracket",
  "close_bracket",
  "backslash",
  "non_us_pound",
  "semicolon",
  "quote",
  "grave_accent_and_tilde",
  "comma",
  "period",
  "slash",
  "non_us_backslash"
], Ce = [
  "up_arrow",
  "down_arrow",
  "left_arrow",
  "right_arrow",
  "page_up",
  "page_down",
  "home",
  "end"
], Ee = [
  "a",
  "b",
  "c",
  "d",
  "e",
  "f",
  "g",
  "h",
  "i",
  "j",
  "k",
  "l",
  "m",
  "n",
  "o",
  "p",
  "q",
  "r",
  "s",
  "t",
  "u",
  "v",
  "w",
  "x",
  "y",
  "z"
], Ae = [
  "1",
  "2",
  "3",
  "4",
  "5",
  "6",
  "7",
  "8",
  "9",
  "0"
], Te = [
  "f1",
  "f2",
  "f3",
  "f4",
  "f5",
  "f6",
  "f7",
  "f8",
  "f9",
  "f10",
  "f11",
  "f12",
  "f13",
  "f14",
  "f15",
  "f16",
  "f17",
  "f18",
  "f19",
  "f20"
], Se = [
  "keypad_num_lock",
  "keypad_slash",
  "keypad_asterisk",
  "keypad_hyphen",
  "keypad_plus",
  "keypad_enter",
  "keypad_1",
  "keypad_2",
  "keypad_3",
  "keypad_4",
  "keypad_5",
  "keypad_6",
  "keypad_7",
  "keypad_8",
  "keypad_9",
  "keypad_0",
  "keypad_period",
  "keypad_equal_sign",
  "keypad_comma"
], Ke = [
  "print_screen",
  "scroll_lock",
  "pause",
  "insert",
  "application",
  "help",
  "power"
], Oe = [
  "international1",
  "international3",
  "lang1",
  "lang2"
], $e = ["japanese_eisuu", "japanese_kana"], je = [
  "volume_down",
  // equal to `volume_decrement`
  "volume_up",
  // equal to `volume_increment`
  "mute",
  "volume_decrement",
  "volume_increment"
], X = [
  // Function keys
  "f21",
  "f22",
  "f23",
  "f24",
  // Keys in pc keyboards
  "execute",
  "menu",
  "select",
  "stop",
  "again",
  "undo",
  "cut",
  "copy",
  "paste",
  "find",
  // International keys
  "international2",
  "international4",
  "international5",
  "international6",
  "international7",
  "international8",
  "international9",
  "lang3",
  "lang4",
  "lang5",
  "lang6",
  "lang7",
  "lang8",
  "lang9",
  // Japanese
  "japanese_pc_nfer",
  // PCキーボードの無変換キー
  "japanese_pc_xfer",
  // PCキーボードの変換キー
  "japanese_pc_katakana",
  // PCキーボードのかなキー
  // Others
  "keypad_equal_sign_as400",
  "locking_caps_lock",
  "locking_num_lock",
  "locking_scroll_lock",
  "alternate_erase",
  "sys_req_or_attention",
  "cancel",
  "clear",
  "prior",
  "return",
  // rarely used return (HID usage 0x9e)
  "separator",
  "out",
  "oper",
  "clear_or_again",
  "cr_sel_or_props",
  "ex_sel"
], F = [
  // Disable this key
  "vk_none",
  // Others
  "vk_consumer_brightness_down",
  // equal to `display_brightness_decrement`
  "vk_consumer_brightness_up",
  // equal to `display_brightness_increment`
  "vk_mission_control",
  // equal to `mission_control`
  "vk_launchpad",
  // equal to `launchpad`
  "vk_dashboard",
  // equal to `dashboard`
  "vk_consumer_illumination_down",
  // equal to `illumination_decrement`
  "vk_consumer_illumination_up",
  // equal to `illumination_increment`
  "vk_consumer_previous",
  // equal to `rewind`
  "vk_consumer_play",
  // equal to `play`
  "vk_consumer_next",
  // equal to `fast_forward`
  "display_brightness_decrement",
  "display_brightness_increment",
  "rewind",
  "play_or_pause",
  "fastforward",
  "apple_display_brightness_decrement",
  "apple_display_brightness_increment",
  "dashboard",
  "launchpad",
  "mission_control",
  "apple_top_case_display_brightness_decrement",
  "apple_top_case_display_brightness_increment",
  "illumination_decrement",
  "illumination_increment"
], Pt = [
  // Media controls
  "rewind",
  "play_or_pause",
  "fast_forward",
  "mute",
  "volume_decrement",
  "volume_increment",
  "al_terminal_lock_or_screensaver",
  // Lock key on Magic Keyboard without Touch ID
  "eject",
  "scan_previous_track",
  "scan_next_track",
  // Application Launch keys
  "al_word_processor",
  "al_text_editor",
  "al_spreadsheet",
  "al_presentation_app",
  "al_email_reader",
  "al_calculator",
  "al_local_machine_browser",
  "al_internet_browser",
  "al_dictionary",
  // Others
  "fastforward"
], Vt = [
  // Media controls
  "menu",
  // Touch ID on Magic Keyboard
  "stop",
  "microphone",
  "selection",
  "bass_boost",
  "loudness",
  "bass_increment",
  "bass_decrement",
  // Application Launch keys
  "al_graphics_editor",
  "al_database_app",
  "al_newsreader",
  "al_voicemail",
  "al_contacts_or_address_book",
  "al_Calendar_Or_Schedule",
  "al_task_or_project_manager",
  "al_log_or_journal_or_timecard",
  "al_checkbook_or_finance",
  "al_a_or_v_capture_or_playback",
  "al_lan_or_wan_browser",
  "al_remote_networking_or_isp_connect",
  "al_network_conference",
  "al_network_chat",
  "al_telephony_or_dialer",
  "al_logon",
  "al_logoff",
  "al_logon_or_logoff",
  "al_control_panel",
  "al_command_line_processor_or_run",
  "al_process_or_task_manager",
  "al_select_task_or_application",
  "al_next_task_or_application",
  "al_previous_task_or_application",
  "al_preemptive_halt_task_or_application",
  "al_integrated_help_center",
  "al_documents",
  "al_thesaurus",
  "al_desktop",
  "al_spell_check",
  "al_grammer_check",
  "al_wireless_status",
  "al_keyboard_layout",
  "al_virus_protection",
  "al_encryption",
  "al_screen_saver",
  "al_alarms",
  "al_clock",
  "al_file_browser",
  "al_power_status",
  "al_image_browser",
  "al_audio_browser",
  "al_movie_browser",
  "al_digital_rights_manager",
  "al_digital_wallet",
  "al_instant_messaging",
  "al_oem_feature_browser",
  "al_oem_help",
  "al_online_community",
  "al_entertainment_content_browser",
  "al_online_shopping_browswer",
  "al_smart_card_information_or_help",
  "al_market_monitor_or_finance_browser",
  "al_customized_corporate_news_browser",
  "al_online_activity_browswer",
  "al_research_or_search_browswer",
  "al_audio_player",
  "al_message_status",
  "al_contact_sync",
  "al_navigation",
  "al_contextaware_desktop_assistant",
  // Generic GUI application Control keys
  "ac_home",
  "ac_back",
  "ac_forward",
  "ac_refresh",
  "ac_bookmarks",
  "ac_search",
  "ac_zoom_out",
  "ac_zoom_in",
  // Remote control buttons
  "menu_pick",
  "menu_up",
  "menu_down",
  "menu_left",
  "menu_right",
  "menu_escape",
  "menu_value_increase",
  "menu_value_decrease",
  "data_on_screen",
  "closed_caption",
  "closed_caption_select",
  "vcr_or_tv",
  "broadcast_mode",
  "snapshot",
  "still",
  "picture_in_picture_toggle",
  "picture_in_picture_swap",
  "red_menu_button",
  "green_menu_button",
  "blue_menu_button",
  "yellow_menu_button",
  "aspect",
  "three_dimensional_mode_select"
], Lt = [
  // Media controls
  "display_brightness_decrement",
  "display_brightness_increment",
  "dictation"
], It = [
  "button1",
  "button2",
  "button3",
  "button4",
  "button5",
  "button6",
  "button7",
  "button8",
  "button9",
  "button10",
  "button11",
  "button12",
  "button13",
  "button14",
  "button15",
  "button16",
  "button17",
  "button18",
  "button19",
  "button20",
  "button21",
  "button22",
  "button23",
  "button24",
  "button25",
  "button26",
  "button27",
  "button28",
  "button29",
  "button30",
  "button31",
  "button32"
], q = {
  "⌘": "command",
  "⌥": "option",
  "⌃": "control",
  "⇧": "shift",
  "⇪": "caps_lock"
}, pe = {
  "↑": "up_arrow",
  "↓": "down_arrow",
  "←": "left_arrow",
  "→": "right_arrow",
  "⇞": "page_up",
  "⇟": "page_down",
  "↖︎": "home",
  "↘︎": "end"
}, he = {
  "⏎": "return_or_enter",
  "⎋": "escape",
  "⌫": "delete_or_backspace",
  "⌦": "delete_forward",
  "⇥": "tab",
  "␣": "spacebar",
  "-": "hyphen",
  "=": "equal_sign",
  "[": "open_bracket",
  "]": "close_bracket",
  "\\": "backslash",
  ";": "semicolon",
  "'": "quote",
  "`": "grave_accent_and_tilde",
  ",": "comma",
  ".": "period",
  "/": "slash"
}, re = {
  ...pe,
  ...he,
  "⇪": q["⇪"]
}, De = [
  ...G,
  ...ve,
  ...Me,
  ...Ce,
  ...Ee,
  ...Ae,
  ...Te,
  ...Se,
  ...Ke,
  ...Oe,
  ...$e,
  ...je,
  ...X,
  ...F
];
function K(t, e, i) {
  if (typeof t == "number") return `${t}`;
  if (t.length > 1 && Y(t)) {
    let n = te(t);
    if ((n == null ? void 0 : n.length) == 1)
      return n[0];
    throw new Error(`Invalid key ${t}`);
  }
  if (t in re)
    return re[t];
  let o = t;
  if (!De.includes(o))
    throw new Error(`${o} is not valid key_code`);
  if (e != null && e.includes(o))
    throw new Error(`Key ${t} cannot be used ${i || "here"}`);
  return o;
}
let Pe = {
  Meh: ["option", "control", "shift"],
  Hyper: ["command", "option", "control", "shift"],
  SuperHyper: ["command", "option", "control", "shift", "fn"]
}, se = {
  "⌘⇧": ["command", "shift"],
  "⌥⇧": ["option", "shift"],
  "⌃⇧": ["control", "shift"],
  "⌘⌥": ["command", "option"],
  "⌘⌃": ["command", "control"],
  "⌥⌃": ["option", "control"],
  "⌘⌥⌃": ["command", "option", "control"],
  "⌘⌥⇧": ["command", "option", "shift"],
  "⌘⌃⇧": ["command", "control", "shift"],
  "⌥⌃⇧": ["option", "control", "shift"],
  "⌘⌥⌃⇧": ["command", "option", "control", "shift"],
  ...Pe
};
function A(t) {
  if (!t) return;
  if (typeof t == "string")
    return Y(t) ? te(t) : t in q ? [q[t]] : t in se ? se[t] : [t];
  if (Array.isArray(t))
    return Le(t) ? Ie(t) : t.map(Be);
  let e;
  "left" in t ? e = D("left", t.left) : "l" in t && (e = D("left", t.l));
  let i;
  if ("right" in t ? i = D("right", t.right) : "r" in t && (i = D("right", t.r)), !(!(e != null && e.length) && !(i != null && i.length)))
    return [...e || [], ...i || []];
}
let me = /^(left|l|<|‹)([⌘⌥⌃⇧]*)$/, ye = /^(right|r|>|›)([⌘⌥⌃⇧]*)$/;
function B(t, e) {
  if (!(!t && !e))
    return {
      mandatory: ae(t),
      optional: ae(e)
    };
}
function ae(t) {
  if (t)
    return t == "any" ? ["any"] : A(t);
}
let Ve = /* @__PURE__ */ new Set([
  "command",
  "option",
  "control",
  "shift"
]);
function Y(t) {
  return me.test(t) || ye.test(t);
}
function te(t) {
  let e = t.match(me);
  if (e)
    return D("left", e[2]);
  let i = t.match(ye);
  if (i)
    return D("right", i[2]);
}
function Le(t) {
  return t.some(Y);
}
function Ie(t) {
  return t.reduce(
    (e, i) => [
      ...e,
      ...Y(i) ? te(i) || [] : A(i) || []
    ],
    []
  );
}
function D(t, e) {
  var i;
  if (e)
    return (i = A(e)) == null ? void 0 : i.map(
      (o) => Ve.has(o) ? `${t}_${o}` : o
    );
}
function Be(t) {
  return t in q ? q[t] : t;
}
function v(t) {
  return Array.isArray(t) ? t : [t];
}
function g(t, e = 1, i) {
  return new M({ type: "variable_if", name: t, value: e, description: i });
}
function Bt(t, e) {
  var o, n;
  let i;
  if (Array.isArray(t))
    i = t.map(z);
  else if (typeof t == "string" || t instanceof RegExp)
    i = [z(t)];
  else
    return new M({
      type: "frontmost_application_if",
      description: e,
      file_paths: (o = t.file_paths) == null ? void 0 : o.map(z),
      bundle_identifiers: (n = t.bundle_identifiers) == null ? void 0 : n.map(z)
    });
  return new M({
    type: "frontmost_application_if",
    description: e,
    bundle_identifiers: i
  });
}
function Nt(t, e) {
  return new M({
    type: "device_if",
    identifiers: v(t),
    description: e
  });
}
function qt(t, e) {
  return new M({
    type: "device_exists_if",
    identifiers: v(t),
    description: e
  });
}
function Rt(t, e) {
  return new M({
    type: "keyboard_type_if",
    keyboard_types: v(t),
    description: e
  });
}
function xt(t, e) {
  return new M({
    type: "input_source_if",
    input_sources: v(t),
    description: e
  });
}
function Ft(t = !0, e) {
  return new M({ type: "event_changed_if", value: t, description: e });
}
function Ut(t, e) {
  return new M({
    type: "expression_if",
    expression: t,
    description: e
  });
}
const Ht = {
  roleString: "accessibility.focused_ui_element.role_string",
  subroleString: "accessibility.focused_ui_element.subrole_string",
  titleString: "accessibility.focused_ui_element.title_string",
  windowPositionX: "accessibility.focused_ui_element.window_position_x",
  windowPositionY: "accessibility.focused_ui_element.window_position_y",
  windowSizeHeight: "accessibility.focused_ui_element.window_size_height",
  windowSizeWidth: "accessibility.focused_ui_element.window_size_width"
};
let Ne = Re({
  frontmost_application_if: "frontmost_application_unless",
  device_if: "device_unless",
  device_exists_if: "device_exists_unless",
  keyboard_type_if: "keyboard_type_unless",
  input_source_if: "input_source_unless",
  variable_if: "variable_unless",
  expression_if: "expression_unless",
  event_changed_if: "event_changed_unless"
});
class M {
  constructor(e) {
    this.condition = e;
  }
  /** Switch type {condition}_if to {condition}_unless, and vice versa */
  unless() {
    return new M({
      ...this.condition,
      type: Ne[this.condition.type]
    });
  }
  build() {
    return { ...this.condition };
  }
}
function qe(t) {
  return typeof t.build == "function";
}
function $(t) {
  return qe(t) ? t.build() : t;
}
function z(t) {
  return typeof t == "string" ? t : t.toString().slice(1, -1);
}
function Re(t) {
  return Object.keys(t).reduce(
    (e, i) => ({ ...e, [e[i]]: i }),
    t
  );
}
let xe = ["optionalAny", "?any", "??"];
function Fe(t) {
  return !t || typeof t != "string" ? !1 : xe.includes(t);
}
function Ue(t) {
  return /^\?(left|l|<|‹|right|r|>|›)?([⌘⌥⌃⇧⇪]*)$/.test(t);
}
function w(t, e) {
  if (!t)
    return B(t, e);
  if (Fe(t))
    return B("", "any");
  if (typeof t == "string") {
    if (Ue(t))
      return B(
        "",
        A(t.slice(1))
      );
    if (t.startsWith("?"))
      throw new Error(`${t} is not valid optional alias`);
  }
  return typeof t == "object" && "optional" in t ? B("", t.optional) : B(t, e);
}
function R(t, e, i) {
  if (typeof t == "object")
    return new V(t);
  let o = K(
    t,
    F,
    "for from.key_code"
  );
  return new V({
    key_code: o,
    modifiers: w(e, i)
  });
}
function zt(t, e, i) {
  return new V({
    consumer_key_code: t,
    modifiers: w(e, i)
  });
}
function Jt(t, e, i) {
  return new V({
    pointing_button: t,
    modifiers: w(e, i)
  });
}
function b(t, e, i) {
  let o = K(
    t,
    X,
    "as to.key_code"
  );
  return {
    ...i,
    key_code: o,
    modifiers: e ? A(e) : void 0
  };
}
function He(t) {
  return b("left_command", "⌥⌃⇧", t);
}
function ze(t) {
  return b("left_option", "⌃⇧", t);
}
function Je(t) {
  return b("fn", "⌘⌥⌃⇧", t);
}
function We(t) {
  return b("vk_none", void 0, t);
}
function Ge(t, e, i) {
  return {
    ...i,
    consumer_key_code: t,
    modifiers: e ? A(e) : void 0
  };
}
function Xe(t, e, i) {
  return {
    ...i,
    generic_desktop: t,
    modifiers: e ? A(e) : void 0
  };
}
function Ye(t, e, i) {
  return {
    ...i,
    pointing_button: t,
    modifiers: e ? A(e) : void 0
  };
}
function Q(t) {
  return { shell_command: t };
}
function Qe(t) {
  let e = t.match(/^"?(.*?)(.app)?"?$/);
  return Q(`open -a "${(e == null ? void 0 : e[1]) || t}".app`);
}
function Ze(t) {
  return Q(`osascript -e '
set prev to the clipboard
set the clipboard to "${t}"
tell application "System Events"
  keystroke "v" using command down
  delay 0.1
end tell
set the clipboard to prev'`);
}
let et = [
  "Tink",
  "Submarine",
  "Sosumi",
  "Morse",
  "Ping",
  "Pop",
  "Purr",
  "Glass",
  "Hero",
  "Frog",
  "Funk",
  "Blow",
  "Bottle",
  "Basso"
];
function Wt(t) {
  let e = et.includes(t) ? `/System/Library/Sounds/${t}.aiff` : t;
  return Q(`afplay ${e}`);
}
function tt(t) {
  return { select_input_source: t };
}
function c(t, e = 1, i, o) {
  return { set_variable: { name: t, value: e, key_up_value: i, type: o } };
}
function it(t, e) {
  return {
    set_variable: {
      name: t,
      expression: e.expression,
      key_up_expression: e.key_up_expression
    }
  };
}
function ot(t) {
  return { set_variable: { name: t, type: "unset" } };
}
function ie(t, e) {
  return { set_notification_message: { id: t, text: e } };
}
function x(t) {
  return { set_notification_message: { id: t, text: "" } };
}
function nt(t) {
  return { mouse_key: t };
}
function rt(t, e = "toggle") {
  return { sticky_modifier: { [t]: e } };
}
function st(t) {
  return {
    software_function: { cg_event_double_click: { button: t } }
  };
}
function at(t) {
  return { software_function: { set_mouse_cursor_position: t } };
}
function lt(t) {
  return {
    software_function: {
      iokit_power_management_sleep_system: { delay_milliseconds: t }
    }
  };
}
function ut(t, e) {
  return { send_user_command: { payload: t, ...e && { endpoint: e } } };
}
function dt() {
  return { from_event: !0 };
}
let P = {};
for (let t = 0; t <= 9; t++)
  P[t] = b(t);
let _t = "abcdefghijklmnopqrstuvwxyz".split("");
for (let t of _t)
  P[t] = b(t), P[t.toUpperCase()] = b(t, "⇧");
let ft = [
  ...Object.keys(pe),
  ...Object.keys(he)
];
for (let t of ft)
  P[t] = b(t);
let ct = {
  1: "!",
  2: "@",
  3: "#",
  4: "$",
  5: "%",
  6: "^",
  7: "&",
  8: "*",
  9: "(",
  0: ")",
  "'": '"',
  "=": "+",
  ";": ":",
  ",": "<",
  ".": ">",
  "/": "?",
  "-": "_",
  "[": "{",
  "\\": "|",
  "]": "}",
  "`": "~"
};
for (let [t, e] of Object.entries(ct))
  P[e] = b(t, "⇧");
function pt(t, e) {
  return t.split("").map((i) => {
    let o = (e == null ? void 0 : e[i]) || P[i];
    if (!o)
      throw new Error(
        `${i} is unknown. Please provide a map with { ${i}: {ToEvent} }`
      );
    return o;
  });
}
class V {
  constructor(e) {
    this.manipulator = { type: "basic", from: e };
  }
  /** The FromEvent of the Manipulator. */
  get from() {
    return this.manipulator.from;
  }
  to(e, i, o) {
    return this.addToEvent(
      typeof e == "object" ? e : b(e, i, o)
    ), this;
  }
  /** To Hyper key ⌘⌥⌃⇧ */
  toHyper(e) {
    return this.addToEvent(He(e)), this;
  }
  /** To Meh key ⌥⌃⇧ */
  toMeh(e) {
    return this.addToEvent(ze(e)), this;
  }
  /** To SuperHyper key ⌘⌥⌃⇧fn */
  toSuperHyper(e) {
    return this.addToEvent(Je(e)), this;
  }
  /** To vk_none (Disable this key) */
  toNone(e) {
    return this.addToEvent(We(e)), this;
  }
  /** To { consumer_key_code } */
  toConsumerKey(e, i, o) {
    return this.addToEvent(Ge(e, i, o)), this;
  }
  /** To { generic_desktop } */
  toGenericDesktop(e, i, o) {
    return this.addToEvent(Xe(e, i, o)), this;
  }
  /** Map to mouse button */
  toPointingButton(e, i, o) {
    return this.addToEvent(Ye(e, i, o)), this;
  }
  /** Map to shell command */
  to$(e) {
    return this.addToEvent(Q(e)), this;
  }
  /** Map to `$ open -a {app}.app` */
  toApp(e) {
    return this.addToEvent(Qe(e)), this;
  }
  /** Map to paste {text} via clipboard */
  toPaste(e) {
    return this.addToEvent(Ze(e)), this;
  }
  /** To change the current input source */
  toInputSource(e) {
    return this.addToEvent(tt(e)), this;
  }
  /** Map to setting a variable */
  toVar(e, i = 1, o, n) {
    return this.addToEvent(c(e, i, o, n)), this;
  }
  /** Map to setting a variable using expressions */
  toVarExpression(e, i) {
    return this.addToEvent(it(e, i)), this;
  }
  /** Map to unsetting a variable */
  toUnsetVar(e) {
    return this.addToEvent(ot(e)), this;
  }
  /** To set or remove (set text to '') the notification message */
  toNotificationMessage(e, i) {
    return this.addToEvent(ie(e, i)), this;
  }
  /** To remove the notification message */
  toRemoveNotificationMessage(e) {
    return this.addToEvent(x(e)), this;
  }
  /** Move mouse cursor by delta */
  toMouseKey(e) {
    return this.addToEvent(nt(e)), this;
  }
  /** Changes to a sticky modifier key */
  toStickyModifier(e, i = "toggle") {
    return this.addToEvent(rt(e, i)), this;
  }
  /** @see https://karabiner-elements.pqrs.org/docs/json/complex-modifications-manipulator-definition/to/software_function/cg_event_double_click/ */
  toCgEventDoubleClick(e) {
    return this.addToEvent(st(e)), this;
  }
  /** Set mouse cursor position */
  toMouseCursorPosition(e) {
    return this.addToEvent(at(e)), this;
  }
  /** To causes a system sleep */
  toSleepSystem(e) {
    return this.addToEvent(lt(e)), this;
  }
  /** Map to send_user_command */
  toSendUserCommand(e, i) {
    return this.addToEvent(ut(e, i)), this;
  }
  /** Pass through the from event as-is */
  toFromEvent() {
    return this.addToEvent(dt()), this;
  }
  /** To type a string of keys */
  toTypeSequence(e, i) {
    return this.addToEvent(pt(e, i)), this;
  }
  toIfAlone(e, i, o) {
    return this.pushOrCreateList(
      this.manipulator,
      "to_if_alone",
      typeof e == "object" ? e : b(e, i, o)
    ), this;
  }
  toIfHeldDown(e, i, o) {
    return this.pushOrCreateList(
      this.manipulator,
      "to_if_held_down",
      typeof e == "object" ? e : b(e, i, o)
    ), this;
  }
  toAfterKeyUp(e, i, o) {
    return this.pushOrCreateList(
      this.manipulator,
      "to_after_key_up",
      typeof e == "object" ? e : b(e, i, o)
    ), this;
  }
  toDelayedAction(e, i) {
    let o = this.manipulator.to_delayed_action || {
      to_if_invoked: [],
      to_if_canceled: []
    };
    return v(e).forEach((n) => o.to_if_invoked.push(n)), v(i).forEach((n) => o.to_if_canceled.push(n)), this.manipulator.to_delayed_action = o, this;
  }
  toIfOtherKeyPressed(e, i, o, n) {
    let r = {
      other_keys: e,
      to: v(
        typeof i == "object" ? i : b(i, o, n)
      )
    };
    return this.pushOrCreateList(this.manipulator, "to_if_other_key_pressed", r), this;
  }
  description(e) {
    return this.manipulator.description = e, this;
  }
  condition(...e) {
    let { conditions: i = [] } = this.manipulator;
    return this.manipulator.conditions = [...i, ...e.map($)], this;
  }
  parameters(e) {
    return this.manipulator.parameters = { ...this.manipulator.parameters, ...e }, this;
  }
  build(e) {
    return [{ ...this.manipulator }];
  }
  addToEvent(e) {
    this.pushOrCreateList(this.manipulator, "to", e);
  }
  pushOrCreateList(e, i, o) {
    let n = e[i] || [];
    v(o).forEach((r) => n.push(r)), Object.assign(e, { [i]: n });
  }
}
function ht(t) {
  return typeof t.build == "function";
}
function O(t, e) {
  return ht(t) ? t.build(e) : "type" in t ? [t] : Object.entries(t).reduce(
    (o, [n, r]) => [...o, ...R(n).to(r).build(e)],
    []
  );
}
function Gt(...t) {
  return (e) => {
    let i = t.map($);
    function o(a) {
      return a.type != "basic" ? a : {
        ...a,
        conditions: [...a.conditions || [], ...i]
      };
    }
    let r = (Array.isArray(e) ? e : O(e)).reduce(
      (a, u) => [...a, ...O(u).map(o)],
      []
    );
    return Object.assign(r, { build: () => r });
  };
}
function Xt(t) {
  return (e) => ({
    build: () => (Array.isArray(t) ? t.map((o, n) => [o, n]) : Object.entries(t)).reduce(
      (o, [n, r]) => [...o, ...O(e(n, r))],
      []
    )
  });
}
function Yt(t, e) {
  return (i) => ({
    build: (o) => {
      let n = w(
        t,
        e
      );
      return (Array.isArray(i) ? i : O(i)).map((a) => O(a, o)).reduce((a, u) => a.concat(u), []).map((a) => {
        var l;
        let u = Object.keys(n).reduce(
          (d, _) => {
            var h;
            if (!((h = n[_]) != null && h.length)) return d;
            let f = /* @__PURE__ */ new Set([
              ...d[_] || [],
              ...n[_]
            ]);
            return d[_] = f.has("any") ? ["any"] : Array.from(f), d;
          },
          { ...(l = a.from) == null ? void 0 : l.modifiers }
        );
        return a.from = { ...a.from, modifiers: u }, a;
      });
    }
  });
}
function mt(t, e, i) {
  let o = (r) => K(r, F, "for from.simultaneous"), n = new yt({
    simultaneous: t.map(
      (r) => typeof r == "object" ? r : { key_code: o(r) }
    ),
    simultaneous_options: e
  });
  return i && n.parameters({
    "basic.simultaneous_threshold_milliseconds": i
  }), n;
}
class yt extends V {
  constructor(e) {
    super(e);
  }
  modifiers(e, i) {
    return this.manipulator.from.modifiers = e || i ? w(e, i) : void 0, this;
  }
}
let be = "__support__manipulator";
function bt(t) {
  return t.description = (t.description || "") + be, t;
}
function oe(t) {
  var e;
  return (e = t.description) == null ? void 0 : e.endsWith(be);
}
let le = {
  "double_tap.delay_milliseconds": 200
};
function Qt(t, e, i, o) {
  let n = K(
    t,
    [...X, ...F],
    "for double tap"
  ), r = new gt({ key_code: n });
  return o ? (r.delay(o), r.from.modifiers = w(
    e,
    i
  )) : i ? typeof i == "number" ? (r.delay(i), r.from.modifiers = w(
    e
  )) : r.from.modifiers = w(
    e,
    i
  ) : e && (typeof e == "number" ? r.delay(e) : r.from.modifiers = w(
    e
  )), r;
}
class gt extends V {
  constructor(e) {
    super(e), this.singleTapEvent = void 0;
  }
  /** Set single tap which is from key by default; null to disable */
  singleTap(e) {
    return this.singleTapEvent = e, this;
  }
  /** Set the delay parameter */
  delay(e) {
    return this.delayParam = e, this;
  }
  build(e) {
    var h, p;
    let i = (e == null ? void 0 : e.getParameters(le)) ?? le, o = this.delayParam || i["double_tap.delay_milliseconds"], n = this.from.key_code;
    if (typeof this.singleTapEvent > "u") {
      this.singleTapEvent = { key_code: n };
      let s = (h = this.manipulator.from.modifiers) == null ? void 0 : h.mandatory;
      s && (this.singleTapEvent.modifiers = s.filter(
        (m) => m != "any"
      ));
    }
    let r = this.singleTapEvent && "key_code" in this.singleTapEvent ? G.includes(this.singleTapEvent.key_code) : !1, a = ["double-tap", n];
    this.from.modifiers && [this.from.modifiers.mandatory, this.from.modifiers.optional].map((s) => s != null && s.length ? s.join(",") : "").forEach((s) => s && a.push(s));
    let u = a.join("-"), l = g(u).build(), d = g(u).unless().build(), _ = bt({
      ...this.manipulator,
      conditions: [l]
    }), f = {
      ...this.manipulator,
      to: [
        c(u, 1),
        ...r ? [
          {
            ...this.singleTapEvent,
            lazy: this.singleTapEvent.lazy || typeof this.singleTapEvent.lazy > "u"
          }
        ] : []
      ],
      conditions: [...this.manipulator.conditions || [], d],
      to_delayed_action: {
        to_if_invoked: [
          ...this.singleTapEvent ? [this.singleTapEvent] : [],
          c(u, 0)
        ],
        to_if_canceled: [c(u, 0)]
      }
    };
    return r && !((p = f.to_if_held_down) != null && p.length) && (f.to_if_held_down = [this.singleTapEvent]), f.parameters = {
      "basic.to_if_held_down_threshold_milliseconds": o,
      "basic.to_delayed_action_delay_milliseconds": o,
      ...this.manipulator.parameters
    }, [_, f];
  }
}
function Zt() {
  return new wt();
}
class wt {
  constructor() {
    this.manipulator = {
      type: "mouse_motion_to_scroll"
    };
  }
  modifiers(e, i) {
    return this.manipulator.from = {
      modifiers: w(
        e,
        i
      )
    }, this;
  }
  condition(...e) {
    let { conditions: i = [] } = this.manipulator;
    return this.manipulator.conditions = [...i, ...e.map($)], this;
  }
  options(e) {
    return this.manipulator.options = { ...this.manipulator.options, ...e }, this;
  }
  build(e) {
    return [{ ...this.manipulator }];
  }
}
function ei(t, ...e) {
  return new Z(t, ...e);
}
class Z {
  constructor(e, ...i) {
    this.ruleDescription = e, this.manipulatorSources = [], this.allowEmptyManipulators = !1, this.conditions = i;
  }
  manipulators(e) {
    return Array.isArray(e) ? e.forEach((i) => this.manipulatorSources.push(i)) : this.manipulatorSources.push(e), this;
  }
  condition(...e) {
    return e.forEach((i) => this.conditions.push(i)), this;
  }
  description(e) {
    return this.ruleDescription = e, this;
  }
  build(e) {
    let i = {
      description: this.ruleDescription,
      manipulators: this.manipulatorSources.reduce(
        (n, r) => [...n, ...O(r, e)],
        []
      )
    };
    if (!this.allowEmptyManipulators && i.manipulators.length == 0)
      throw new Error(`"manipulators" is empty in "${i.description}"`);
    if (this.conditions.length == 0) return i;
    let o = this.conditions.map($);
    return i.manipulators = i.manipulators.map(
      (n) => n.type == "basic" && !oe(n) ? { ...n, conditions: [...n.conditions || [], ...o] } : { ...n }
    ), i;
  }
}
function kt(t) {
  return typeof t.build == "function";
}
function vt(t, e) {
  return kt(t) ? t.build(e) : t;
}
let W = {
  escape: ["escape", "caps_lock"]
};
function ge(t, e, i) {
  let o = [];
  if (!t) return o;
  for (let n of Array.isArray(t) ? t : [t]) {
    let r = R(n);
    o.push(...r.condition(e).to(i).build());
  }
  return o;
}
let ue = {
  "layer.delay_by_default": !1,
  "layer.delay_milliseconds": 200
}, ee = [
  ...F,
  ...X,
  ...G
];
function Mt(t, e, i = 1, o = 0) {
  return new Et(t, e, i, o);
}
function Ct(t, e, i, o = 1, n = 0) {
  return Mt(e, i, o, n).modifiers(t);
}
function ti(t, e, i = 1, o = 0) {
  return Ct("Hyper", t, e, i, o);
}
let J = "__layer", de = "__delay";
class Et extends Z {
  constructor(e, i, o = 1, n = 0) {
    let r = v(e).map(
      (a) => K(a, ee, "as layer key")
    );
    i || (i = `layer-${r.join("-")}`), super(`Layer - ${i}`), this.onValue = o, this.offValue = n, this.replaceLayerKeyToIfAlone = !1, this.delayed = void 0, this.keys = r, this.varName = i, this.layerCondition = g(this.varName, this.onValue), this.condition(this.layerCondition), this.allowEmptyManipulators = !0;
  }
  modifiers(e, i) {
    return this.layerModifiers = e || i ? w(e, i) : void 0, this;
  }
  /** Config the layer key. */
  configKey(e, i = !1) {
    return this.layerKeyManipulator || (this.layerKeyManipulator = R("fn")), e(this.layerKeyManipulator), this.replaceLayerKeyToIfAlone = i, this;
  }
  /** Set the notification when the layer is active. */
  notification(e = !0) {
    return this.layerNotification = e, this;
  }
  /** Set leader mode. Default escape keys: ['escape', 'caps_lock']. */
  leaderMode(e = !0) {
    return typeof e == "boolean" ? this.leaderModeOptions = e ? W : void 0 : this.leaderModeOptions = { ...W, ...e }, this;
  }
  /** Set delay mode (and delay_milliseconds). */
  delay(e = !0) {
    return this.delayed = e, this;
  }
  build(e) {
    var n, r, a, u;
    let i = super.build(e);
    if (this.leaderModeOptions) {
      let l = [
        c(this.varName, this.offValue),
        c(J, 0)
      ];
      this.layerNotification && l.push(x(ke(this.varName))), this.leaderModeOptions.sticky || i.manipulators.forEach(
        (d) => d.type == "basic" && !oe(d) && (d.to = (d.to || []).concat(l))
      ), i.manipulators.push(
        ...ge(
          this.leaderModeOptions.escape,
          g(this.varName, this.onValue),
          l
        )
      );
    }
    if ((r = (n = this.layerModifiers) == null ? void 0 : n.mandatory) != null && r.length || (u = (a = this.layerModifiers) == null ? void 0 : a.optional) != null && u.length) {
      let l = _e(this.layerModifiers) == "optional";
      i.manipulators.forEach(
        (d) => this.addModifierAnyToManipulator(d, l)
      );
    }
    let o = this.conditions.filter((l) => l != this.layerCondition).map($);
    for (let l of this.keys)
      i.manipulators = [
        ...we(
          l,
          this.varName,
          this.onValue,
          this.offValue,
          this.layerModifiers,
          o,
          e,
          this.layerKeyManipulator,
          this.replaceLayerKeyToIfAlone,
          typeof this.layerNotification == "boolean" && this.layerNotification ? this.ruleDescription : this.layerNotification || void 0,
          this.leaderModeOptions,
          this.delayed
        ),
        ...i.manipulators
      ];
    return i;
  }
  // If the layer has modifiers, set manipulator modifiers to { mandatory: ['any'] }
  addModifierAnyToManipulator(e, i) {
    if (e.type == "basic") {
      if (e.from.modifiers) {
        let { mandatory: o, optional: n } = e.from.modifiers;
        if (n != null && n.length || o != null && o.length) {
          let r = _e(e.from.modifiers);
          if (r == "mandatory")
            e.from.modifiers = { mandatory: ["any"] };
          else if (r == "optional")
            e.from.modifiers = { optional: ["any"] };
          else if (!i)
            throw new Error(
              "Layers with modifiers cannot have modifiers on manipulators"
            );
          return;
        }
      }
      e.from.modifiers = i ? { optional: ["any"] } : { mandatory: ["any"] };
    }
  }
}
function we(t, e, i, o, n, r, a, u, l, d, _, f) {
  var I;
  function h(y) {
    var ne;
    if (!u) return y;
    let j = u.build()[0], S = v(y)[0];
    if ([
      "to",
      "to_if_alone",
      "to_if_held_down",
      "to_after_key_up"
    ].forEach(
      (C) => {
        var U;
        return (U = j[C]) == null ? void 0 : U.forEach(
          (H) => S[C] = [...S[C] || [], H]
        );
      }
    ), j.to_delayed_action) {
      S.to_delayed_action = S.to_delayed_action || {
        to_if_invoked: [],
        to_if_canceled: []
      };
      for (let C of ["to_if_invoked", "to_if_canceled"])
        j.to_delayed_action[C].forEach(
          (U) => {
            var H;
            return (H = S.to_delayed_action) == null ? void 0 : H[C].push(U);
          }
        );
    }
    return l && (S.to_if_alone = (ne = S.to_if_alone) == null ? void 0 : ne.filter(
      (C) => !("key_code" in C && C.key_code == t)
    )), y;
  }
  let p = At(f, a), s = R({ key_code: t, modifiers: n }).toVar(e, p > 0 ? de : i).toVar(J).condition(
    p > 0 ? g(e, o) : g(e, i).unless(),
    g(J).unless()
  );
  !((I = n == null ? void 0 : n.mandatory) != null && I.length) && !_ && s.toIfAlone({ key_code: t }), _ || s.toAfterKeyUp(c(e, o)).toAfterKeyUp(c(J, 0));
  let m = g(e, de).build();
  if (p > 0 && s.toIfAlone(c(e, o)).toIfHeldDown(c(e, i)).toDelayedAction(
    [],
    [
      { key_code: t, conditions: [m] },
      {
        set_variable: { name: e, value: o },
        conditions: [m]
      }
    ]
  ).parameters({
    "basic.to_if_held_down_threshold_milliseconds": p,
    "basic.to_delayed_action_delay_milliseconds": p,
    "basic.to_if_alone_timeout_milliseconds": p
  }), r != null && r.length && s.condition(...r), d) {
    let y = ke(e);
    p > 0 ? (s.toIfHeldDown(ie(y, d)), s.toDelayedAction([], {
      ...x(y),
      conditions: [m]
    })) : s.toNotificationMessage(y, d), _ || s.toAfterKeyUp(x(y));
  }
  if (!a)
    return h(s.build());
  let T = [
    `layer_${t}`,
    ...n ? [JSON.stringify(n)] : [],
    ...(r || []).map((y) => JSON.stringify(y)).sort()
  ].join("_"), k = a.getCache(T);
  if (k != null && k.to && k.to_after_key_up)
    return k.to.find(
      (j) => "set_variable" in j && j.set_variable.name == e
    ) || (k.to.push(c(e, i)), k.to_after_key_up.push(c(e, o))), h(k), [];
  let L = s.build(a);
  return a.setCache(T, L[0]), h(L);
}
function _e({
  mandatory: t,
  optional: e
}) {
  return (t == null ? void 0 : t.length) == 1 && t[0] == "any" ? "mandatory" : (e == null ? void 0 : e.length) == 1 && e[0] == "any" ? "optional" : null;
}
function ke(t) {
  return `layer-${t}`;
}
function At(t, e) {
  if (typeof t == "number") return t;
  if (t == !1) return 0;
  let i = (e == null ? void 0 : e.getParameters(ue)) ?? ue;
  return t || i["layer.delay_by_default"] ? i["layer.delay_milliseconds"] : 0;
}
let fe = {
  "simlayer.threshold_milliseconds": 200
};
function ii(t, e, i, o = 1, n = 0) {
  return new Tt(t, e, i, o, n);
}
class Tt extends Z {
  constructor(e, i, o, n = 1, r = 0) {
    let a = v(e).map(
      (u) => K(u, ee, "as simlayer key")
    );
    i || (i = `simlayer-${a.join("-")}`), super(`Simlayer - ${i}`), this.threshold = o, this.onValue = n, this.offValue = r, this.sharedLayerKeys = [], this.simultaneousOptions = {
      detect_key_down_uninterruptedly: !0,
      key_down_order: "strict",
      key_up_order: "strict_inverse",
      key_up_when: "any"
    }, this.layerModifiers = { optional: ["any"] }, this.ifActivated = [], this.ifDeactivated = [], this.keys = a, this.varName = i, this.layerCondition = g(this.varName, this.onValue), this.condition(this.layerCondition);
  }
  modifiers(e, i) {
    return this.layerModifiers = e || i ? w(e, i) : void 0, this;
  }
  /** Set simultaneous_options on the simlayer toggle manipulator */
  options(e) {
    return Object.assign(this.simultaneousOptions, e), this;
  }
  /** Enable layer with the same variable and manipulators with this simlayer */
  enableLayer(...e) {
    return e.map(
      (i) => K(i, ee, "as layer key")
    ).forEach((i) => {
      if (this.keys.includes(i))
        throw new Error(`Key ${i} is already used in ${this.ruleDescription}`);
      if (this.sharedLayerKeys.includes(i))
        throw new Error(
          `Key ${i} is already used as shared layer key in  ${this.ruleDescription}`
        );
      this.sharedLayerKeys.push(i);
    }), this;
  }
  /** The ToEvents to trigger when the layer is activated */
  toIfActivated(e) {
    return this.ifActivated.push(e), this;
  }
  /** The ToEvents to trigger when the layer is deactivated */
  toIfDeactivated(e) {
    return this.ifDeactivated.push(e), this;
  }
  build(e) {
    let i = super.build(e), o = (e == null ? void 0 : e.getParameters(fe)) ?? fe, n = this.threshold || o["simlayer.threshold_milliseconds"], r = this.conditions.length > 1 ? this.conditions.filter((l) => l != this.layerCondition).map($) : void 0, a = c(this.varName, this.onValue), u = c(this.varName, this.offValue);
    i.manipulators.concat().forEach((l) => {
      var _;
      if (l.type != "basic")
        throw new Error(
          `Unsupported manipulator type ${l.type} in simlayer ${this.ruleDescription}`
        );
      let d = (_ = l.from) == null ? void 0 : _.key_code;
      if (!d)
        throw new Error(
          `Missing from.key_code in simlayer ${this.ruleDescription}`
        );
      this.layerModifiers && (l.from.modifiers = {
        ...l.from.modifiers,
        ...this.layerModifiers
      });
      for (let f of this.keys)
        i.manipulators.push({
          type: "basic",
          parameters: {
            "basic.simultaneous_threshold_milliseconds": n
          },
          to: [a, ...l.to || [], ...this.ifActivated],
          from: {
            simultaneous: [{ key_code: f }, { key_code: d }],
            simultaneous_options: {
              ...this.simultaneousOptions,
              to_after_key_up: [
                ...this.simultaneousOptions.to_after_key_up || [],
                u,
                ...this.ifDeactivated
              ]
            },
            modifiers: this.layerModifiers
          },
          conditions: r
        });
    });
    for (let l of this.sharedLayerKeys)
      i.manipulators = [
        ...we(
          l,
          this.varName,
          this.onValue,
          this.offValue,
          this.layerModifiers,
          r,
          e
        ),
        ...i.manipulators
      ];
    return i;
  }
}
let ce = {
  "duo_layer.threshold_milliseconds": 200,
  "duo_layer.notification": !1,
  "duo_layer.delay_by_default": !1,
  "duo_layer.delay_milliseconds": 200
};
function oi(t, e, i, o = 1, n = 0) {
  return new St(t, e, i, o, n);
}
class St extends Z {
  constructor(e, i, o, n = 1, r = 0) {
    let a = `DuoLayer ${o || `${e} ${i}`}`;
    o || (o = `duo-layer-${e}-${i}`), super(a), this.key1 = e, this.key2 = i, this.onValue = n, this.offValue = r, this.simultaneousOptions = {}, this.ifActivated = [], this.ifDeactivated = [], this.delayed = void 0, this.varName = o, this.layerCondition = g(this.varName, this.onValue), this.condition(this.layerCondition), this.allowEmptyManipulators = !0;
  }
  threshold(e) {
    return this.simultaneousThreshold = e, this;
  }
  options(e) {
    return Object.assign(this.simultaneousOptions, e), this;
  }
  /** Set the notification when the layer is active. */
  notification(e = !0) {
    return this.layerNotification = e, this;
  }
  /** The ToEvents to trigger when the layer is activated */
  toIfActivated(e) {
    return this.ifActivated.push(e), this;
  }
  /** The ToEvents to trigger when the layer is deactivated */
  toIfDeactivated(e) {
    return this.ifDeactivated.push(e), this;
  }
  /** Set leader mode. Default escape keys: ['escape', 'caps_lock']. */
  leaderMode(e = !0) {
    return typeof e == "boolean" ? this.leaderModeOptions = e ? W : void 0 : this.leaderModeOptions = { ...W, ...e }, this;
  }
  /** Set delay mode (and delay_milliseconds). */
  delay(e = !0) {
    return this.delayed = e, this;
  }
  build(e) {
    let i = super.build(e), o = (e == null ? void 0 : e.getParameters(ce)) ?? ce, n = this.simultaneousThreshold || o["duo_layer.threshold_milliseconds"], r = this.layerNotification ?? o["duo_layer.notification"], a = typeof this.delayed == "number" ? this.delayed : this.delayed == !1 ? 0 : this.delayed == !0 || o["duo_layer.delay_by_default"] ? o["duo_layer.delay_milliseconds"] : 0, u = this.conditions.filter((s) => s != this.layerCondition).map($), l = [c(this.varName, this.onValue), ...this.ifActivated], d = [
      c(this.varName, this.offValue),
      ...this.ifDeactivated
    ];
    if (r) {
      let s = `duo-layer-${this.varName}`, m = typeof r == "boolean" ? this.ruleDescription : r;
      l.push(ie(s, m)), d.push(x(s));
    }
    this.leaderModeOptions && (this.leaderModeOptions.sticky || i.manipulators.forEach(
      (s) => s.type == "basic" && !oe(s) && (s.to = (s.to || []).concat(d))
    ), i.manipulators.push(
      ...ge(
        this.leaderModeOptions.escape,
        g(this.varName, this.onValue),
        d
      )
    ));
    let _ = this.simultaneousOptions.to_after_key_up || [];
    this.leaderModeOptions || _.push(...d);
    let f = [
      [this.key1, this.key2],
      ...a > 0 ? [[this.key2, this.key1]] : []
    ].map((s) => {
      let m = mt(
        s,
        {
          ...this.simultaneousOptions,
          ...a > 0 ? { key_down_order: "strict" } : {},
          to_after_key_up: _
        },
        n
      ).modifiers("??").condition(g(this.varName, this.onValue).unless());
      return u.length && m.condition(...u), a > 0 ? m.toIfHeldDown(l).toIfAlone(s.map((T) => b(T))).parameters({
        "basic.to_if_held_down_threshold_milliseconds": a,
        "basic.to_if_alone_timeout_milliseconds": a
      }) : m.to(l), m;
    });
    if (!e)
      return i.manipulators = [
        ...f.flatMap((s) => s.build()),
        ...i.manipulators
      ], i;
    let h = [
      "duo_layer",
      ...[this.key1, this.key2].sort(),
      ...u.map((s) => JSON.stringify(s)).sort()
    ].join("_"), p = e.getCache(h);
    if (p)
      p.forEach((s) => {
        var T, k, L, I;
        ((T = s.to) == null ? void 0 : T.find(
          (y) => "set_variable" in y && y.set_variable.name == this.varName
        )) || ((k = s.to) == null || k.push(c(this.varName, this.onValue)), (I = (L = s.from.simultaneous_options) == null ? void 0 : L.to_after_key_up) == null || I.push(
          c(this.varName, this.offValue)
        ));
      });
    else {
      let s = f.flatMap((m) => m.build(e));
      e.setCache(h, s), i.manipulators = [...s, ...i.manipulators];
    }
    return i;
  }
}
class Kt {
  constructor() {
    this.parameters = {}, this.cache = /* @__PURE__ */ new Map();
  }
  setParameters(e) {
    Object.assign(this.parameters, e);
  }
  getParameters(e) {
    let i = { ...e };
    for (let o of Object.keys(e))
      o in this.parameters && this.parameters[o] != null && Object.assign(i, { [o]: this.parameters[o] });
    return i;
  }
  getCache(e) {
    return this.cache.get(e);
  }
  setCache(e, i) {
    this.cache.set(e, i);
  }
}
let Ot = {
  "basic.to_if_alone_timeout_milliseconds": 1e3,
  "basic.to_if_held_down_threshold_milliseconds": 500,
  "basic.to_delayed_action_delay_milliseconds": 500,
  "basic.simultaneous_threshold_milliseconds": 50,
  "mouse_motion_to_scroll.speed": 100
};
function $t(t, e = {}) {
  let {
    "double_tap.delay_milliseconds": i,
    "layer.delay_by_default": o,
    "layer.delay_milliseconds": n,
    "simlayer.threshold_milliseconds": r,
    "duo_layer.threshold_milliseconds": a,
    "duo_layer.notification": u,
    "duo_layer.delay_by_default": l,
    "duo_layer.delay_milliseconds": d,
    ..._
  } = e, f = new Kt();
  f.setParameters({
    "double_tap.delay_milliseconds": i
  }), f.setParameters({
    "layer.delay_by_default": o,
    "layer.delay_milliseconds": n
  }), f.setParameters({
    "simlayer.threshold_milliseconds": r
  }), f.setParameters({
    "duo_layer.threshold_milliseconds": a,
    "duo_layer.notification": u,
    "duo_layer.delay_by_default": l,
    "duo_layer.delay_milliseconds": d
  });
  let h = t.map((s) => vt(s, f)), p = {
    rules: h.filter((s) => s.manipulators.length),
    parameters: {
      ...Ot,
      ..._
    }
  };
  if (p.rules.length == 0)
    throw new Error('complex_modifications "rules" is empty ');
  return p.rules.length < h.length && console.warn(`Rules with empty manipulators are ignored: 
${h.filter((s) => s.manipulators.length == 0).map((s) => "- " + s.description).join(`
`)}
`), p;
}
function jt(t) {
  return t.reduce(
    (o, n) => [...o, ...O(n)],
    []
  ).map((o) => {
    if (!(o.type == "basic" && "from" in o && "to" in o))
      throw new Error(
        "simple_modifications manipulator isn't a BasicManipulator"
      );
    let { to: n, from: r, type: a, ...u } = o;
    if (!n || !r)
      throw new Error("simple_modifications manipulator missing to/from");
    if (Object.keys(u).length)
      throw new Error(
        `simple_modifications manipulator contains unsupported fields: ${Object.keys(u).join(", ")}. Only 'from' and 'to' are allowed.`
      );
    return { to: n, from: r };
  });
}
let E = {
  karabinerConfigDir() {
    return require("node:path").join(
      require("node:os").homedir(),
      ".config/karabiner"
    );
  },
  karabinerConfigFile() {
    return require("node:path").join(
      this.karabinerConfigDir(),
      "karabiner.json"
    );
  },
  readKarabinerConfig(t) {
    return require(t ?? this.karabinerConfigFile());
  },
  writeKarabinerConfig(t, e) {
    return require("node:fs/promises").writeFile(
      e ?? this.karabinerConfigFile(),
      t
    );
  },
  readJson(t) {
    return require(t);
  },
  exit(t = 0) {
    process.exit(t);
  }
};
function ni(t, e, i = {}, o) {
  typeof t == "string" && (t = { name: t, dryRun: t == "--dry-run" });
  let { name: n, dryRun: r } = t, a = t.karabinerJsonPath ?? E.karabinerConfigFile(), u = r ? { profiles: [{ name: n, complex_modifications: { rules: [] } }] } : E.readKarabinerConfig(a), l = u == null ? void 0 : u.profiles.find((_) => _.name == n);
  l || N(`⚠️ Profile ${n} not found in ${a}.

ℹ️ Please check the profile name in the Karabiner-Elements UI and 
    - Update the profile name at writeToProfile()
    - Create a new profile if needed
 `);
  try {
    l.complex_modifications = $t(e, i);
  } catch (_) {
    N(_);
  }
  if (o != null && o.simple_modifications)
    try {
      l.simple_modifications = jt(
        o.simple_modifications
      );
    } catch (_) {
      N(_);
    }
  let d = JSON.stringify(u, null, 2);
  if (r) {
    console.info(d);
    return;
  }
  E.writeKarabinerConfig(d, a).catch(N), console.log(`✓ Profile ${n} updated.`);
}
function N(t) {
  return t && console.error(typeof t == "string" ? t : t.stack || t.message || t), E.exit(1);
}
function ri(t, e) {
  let i = e ?? E.karabinerConfigFile(), o = E.readKarabinerConfig(i);
  o.global = { ...o.global, ...t };
  let n = JSON.stringify(o, null, 2);
  E.writeKarabinerConfig(n, i).catch(N), console.log("✓ global updated.");
}
function si(t) {
  return {
    build() {
      let e = E.readJson(t);
      if (!Array.isArray(e == null ? void 0 : e.rules))
        throw new Error(`Cannot file rules in ${t}`);
      return {
        description: `Imported from ${t}`,
        manipulators: e.rules.reduce(
          (i, o) => i.concat(o.manipulators),
          []
        )
      };
    }
  };
}
function ai(t, e) {
  return {
    build() {
      let o = E.readKarabinerConfig(e).profiles.find((n) => n.name == t);
      if (!o) throw new Error(`Profile ${t} not found`);
      return {
        description: `Imported from profile ${t}`,
        manipulators: o.complex_modifications.rules.reduce(
          (n, r) => n.concat(r.manipulators),
          []
        )
      };
    }
  };
}
export {
  V as BasicManipulatorBuilder,
  M as ConditionBuilder,
  Ht as accessibilityVariable,
  pe as arrowKeyAliases,
  Ce as arrowKeyCodes,
  $ as buildCondition,
  $t as complexModifications,
  he as controlOrSymbolKeyAliases,
  Me as controlOrSymbolKeyCodes,
  Ot as defaultComplexModificationsParameters,
  le as defaultDoubleTapParameters,
  ce as defaultDuoLayerParameters,
  ue as defaultLayerParameters,
  fe as defaultSimlayerParameters,
  oi as duoLayer,
  Pt as fromAndToConsumerKeyCodes,
  Vt as fromOnlyConsumerKeyCodes,
  X as fromOnlyKeyCodes,
  Te as functionKeyCodes,
  K as getKeyWithAlias,
  ti as hyperLayer,
  Bt as ifApp,
  Nt as ifDevice,
  qt as ifDeviceExists,
  Ft as ifEventChanged,
  Ut as ifExpression,
  xt as ifInputSource,
  Rt as ifKeyboardType,
  g as ifVar,
  si as importJson,
  ai as importProfile,
  Oe as internationalKeyCodes,
  qe as isConditionBuilder,
  Y as isSideMultiModifierAlias,
  $e as japaneseKeyCodes,
  Se as keypadKeyCodes,
  Mt as layer,
  Ee as letterKeyCodes,
  R as map,
  zt as mapConsumerKey,
  Qt as mapDoubleTap,
  Jt as mapPointingButton,
  mt as mapSimultaneous,
  q as modifierKeyAliases,
  ve as modifierKeyCodes,
  Ct as modifierLayer,
  Zt as mouseMotionToScroll,
  se as multiModifierAliases,
  Pe as namedMultiModifierAliases,
  Ae as numberKeyCodes,
  je as otherKeyCodes,
  B as parseFromModifierParams,
  A as parseModifierParam,
  te as parseSideMultiModifierAlias,
  Ke as pcKeyboardKeyCodes,
  It as pointingButtons,
  ei as rule,
  ii as simlayer,
  jt as simpleModifications,
  G as stickyModifierKeyCodes,
  et as systemSounds,
  Q as to$,
  Qe as toApp,
  st as toCgEventDoubleClick,
  Ge as toConsumerKey,
  dt as toFromEvent,
  Xe as toGenericDesktop,
  He as toHyper,
  tt as toInputSource,
  b as toKey,
  ze as toMeh,
  at as toMouseCursorPosition,
  nt as toMouseKey,
  We as toNone,
  ie as toNotificationMessage,
  Lt as toOnlyConsumerKeyCodes,
  F as toOnlyKeyCodes,
  Ze as toPaste,
  Wt as toPlaySound,
  Ye as toPointingButton,
  x as toRemoveNotificationMessage,
  ut as toSendUserCommand,
  c as toSetVar,
  it as toSetVarExpression,
  lt as toSleepSystem,
  rt as toStickyModifier,
  Je as toSuperHyper,
  pt as toTypeSequence,
  ot as toUnsetVar,
  Gt as withCondition,
  Xt as withMapper,
  Yt as withModifier,
  ri as writeToGlobal,
  ni as writeToProfile
};
